docs/services/ml/upload-training-script.md

```diff
@@ -0,0 +1,611 @@
+---
+title: "Train a Model with a Custom Python Training Script"
+linkTitle: "Use a Custom Training Script"
+weight: 90
+type: "docs"
+tags: ["data management", "ml", "model training"]
+description: "How to write and upload a custom machine learning model training script to the Viam registry and use it to submit training jobs."
+no_service: true
+# SME: Tahiya S.
+---
+
+You can upload a custom training script to the [Viam registry](https://app.viam.com/registry/) to train your ML model.
+Follow this guide to create, upload, and submit your Python script that loads a training dataset, trains an ML model, and produces a model artifact.
+
+## Create a training script
+
+You must save your training script in the `tar.gz` format to run in the Viam ML training service.
+Create a Python source distribution `tar.gz` file that contains all of the training code for training your model:
+
+1. Include an entrypoint file that contains the main logic at a predetermined location like <file>model/training.py</file>.
+   This script must take as command line inputs a `dataset_file` and `model_output_directory` and output the model artifacts it generates to the `model_output_directory`.
+
+- The `dataset_file` is a file containing the GCS blob paths and annotations associated with the dataset ID on which the user is training.
+- The `model_output_directory` is the location where the user saves the model artifacts (the TFLite, PyTorch, TF, ONNX files) they produce from training.
```

Comment:
What does it mean by save "to Viam" or save "to GCS" here-- I can't tell what's happening from code besides that it looks like its saving the model locally (?) with f.write? @tahiyasalam

Reply:
It's saving the model to the output directory that's parsed in the arguments. Then everything that's written to that directory converted into a package/registry item. That's what it means by save the model to Viam.

Reply:
sorry this might be a stupid question but where is the filesystem that the output directory is expected to be on? like is it the user's computer? @tahiyasalam

Reply:
Not a stupid question. This script runs a training job in the cloud. So the user writes a training scripts, submits it, and then we run it in the cloud. In the backend, we provide a Viam produced dataset file and then a model output directory to the script; these are the arguments that the user needs to parse within their own training script. Then, the user can save the model outputs to this model output directory in the "cloud". (Similarly within the script, the user needs to parse the dataset file and then feed it into the model) Once the job is complete, Viam looks at that directory and creates a package and registry item with all of the contents of the directory. Then, the user can interact with the model artifacts in the registry (e.g. deploy the model). I feel like it would be helpful to have this information in the docs somewhere. Let me know if you have any questions or want to chat. As I've mentioned, happy to talk about this at any level since it's a complex project with a significant dev X component.

Reply:
ok, tried to update with some of this info-- lmk what you think